Deprecated API of libpcap (pcap_lookupdev) used

<!--
If reporting an issue please try to provide the information asked below.

Before reporting an issue please:

1. make sure that you're using the latest published GVM components for the
   release you are using: http://openvas.org/install-source.html.
2. check the list of issues whether it isn't already reported.
3. be aware that this is not a support forum. If your issue is more a question
   than a bug report, please use https://community.greenbone.net/c/gse instead.
-->

### Expected behavior
Compilation without warnings

### Current behavior
Compilation generates warning (in Fedora error) about using deprecated API

### Steps to reproduce
export 'CFLAGS=-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection'
+ CFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection'

CXXFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection'
export CXXFLAGS

FFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -I/usr/lib64/gfortran/modules'
export FFLAGS

FCFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -I/usr/lib64/gfortran/modules'
export FCFLAGS

LDFLAGS='-Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld'
export LDFLAGS

 /usr/bin/cmake -DCMAKE_C_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_Fortran_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -DCMAKE_INSTALL_PREFIX:PATH=/usr -DINCLUDE_INSTALL_DIR:PATH=/usr/include -DLIB_INSTALL_DIR:PATH=/usr/lib64 -DSYSCONF_INSTALL_DIR:PATH=/etc -DSHARE_INSTALL_PREFIX:PATH=/usr/share -DLIB_SUFFIX=64 -DBUILD_SHARED_LIBS:BOOL=ON -DLOCALSTATEDIR:PATH=/var -DOPENVASSD_RULES:PATH=/etc/openvas/openvassd.rules


### GVM versions
**gvm-libs:** 11.0.0
**openvas-scanner:** 7.0.0

### Environment
**Operating system:**
Fedora 30

### Logfiles
```
cd /home/mambroz/rpmbuild/BUILD/openvas-7.0.0/misc && /usr/lib64/ccache/cc -DOPENVAS_DATA_DIR=\"/usr/share/openvas\" -DOPENVAS_STATE_DIR=\"/var/lib/openvas\" -DOPENVAS_SYSCONF_DIR=\"/etc/openvas\" -Dopenvas_misc_shared_EXPORTS -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/gvm -I/usr/include/p11-kit-1  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_FILE_OFFSET_BITS=64 -DLARGEFILE_SOURCE=1 -Wall -D_BSD_SOURCE -D_ISOC99_SOURCE -D_SVID_SOURCE -D_DEFAULT_SOURCE -Wall -Wextra -fPIC -g  -Werror -fPIC   -o CMakeFiles/openvas_misc_shared.dir/plugutils.c.o   -c /home/mambroz/rpmbuild/BUILD/openvas-7.0.0/misc/plugutils.c
/home/mambroz/rpmbuild/BUILD/openvas-7.0.0/misc/bpf_share.c: In function 'bpf_open_live':
/home/mambroz/rpmbuild/BUILD/openvas-7.0.0/misc/bpf_share.c:70:5: error: 'pcap_lookupdev' is deprecated: use 'pcap_findalldevs' and use the first device [-Werror=deprecated-declarations]
   70 |     iface = pcap_lookupdev (errbuf);
      |     ^~~~~
In file included from /usr/include/pcap.h:43,
                 from /home/mambroz/rpmbuild/BUILD/openvas-7.0.0/misc/bpf_share.c:27:
/usr/include/pcap/pcap.h:326:16: note: declared here
  326 | PCAP_API char *pcap_lookupdev(char *)
      |                ^~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,6 +10,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - The logging of the NASL internal regexp functions was extended to include the pattern in case of a failed regcomp(). [#397](https://github.com/greenbone/openvas/pull/397)
 - Add config for gpg keyring path (OPENVAS_GPG_BASE_DIR) [#407](https://github.com/greenbone/openvas/pull/407)
 - Use __func__ instead of __FUNCTION__ [#419](https://github.com/greenbone/openvas/pull/419)
+- Use pcap_findalldevs() instead of deprecated function pcap_lookupdev() [#422](https://github.com/greenbone/openvas/pull/422)
 
 [Unreleased]: https://github.com/greenbone/openvas/compare/openvas-7.0...master
 

--- a/misc/bpf_share.c
+++ b/misc/bpf_share.c
@@ -45,6 +45,7 @@ print_pcap_error (pcap_t *p, char *prefix)
 }
 
 /**
+ * @param iface Name of interface. String has to be freed by caller.
  * @return -1 in case of error, index of the opened pcap_t in pcaps
  *         otherwise.
  */
@@ -53,6 +54,7 @@ bpf_open_live (char *iface, char *filter)
 {
   char errbuf[PCAP_ERRBUF_SIZE];
   pcap_t *ret;
+  pcap_if_t *alldevsp = NULL; /* list of capture devices */
   bpf_u_int32 netmask, network;
   struct bpf_program filter_prog;
   int i;
@@ -67,7 +69,14 @@ bpf_open_live (char *iface, char *filter)
     }
 
   if (iface == NULL)
-    iface = pcap_lookupdev (errbuf);
+    {
+      if (pcap_findalldevs (&alldevsp, errbuf) == -1)
+        g_message ("Error for pcap_findalldevs(): %s", errbuf);
+      if (alldevsp != NULL)
+        /* get first device in list */
+        iface = g_strdup (alldevsp->name);
+      pcap_freealldevs (alldevsp);
+    }
 
   ret = pcap_open_live (iface, 1500, 0, 1, errbuf);
   if (ret == NULL)

--- a/nasl/capture_packet.c
+++ b/nasl/capture_packet.c
@@ -48,6 +48,7 @@ init_capture_device (struct in_addr src, struct in_addr dest, char *filter)
   char *a_dst, *a_src;
   char errbuf[PCAP_ERRBUF_SIZE];
   int free_filter = 0;
+  pcap_if_t *alldevsp = NULL; /* list of capture devices */
 
   a_src = g_strdup (inet_ntoa (src));
   a_dst = g_strdup (inet_ntoa (dest));
@@ -72,12 +73,25 @@ init_capture_device (struct in_addr src, struct in_addr dest, char *filter)
   g_free (a_dst);
   g_free (a_src);
 
-  if ((interface = routethrough (&src, &dest))
-      || (interface = pcap_lookupdev (errbuf)))
-    ret = bpf_open_live (interface, filter);
+  if ((interface = routethrough (&src, &dest)))
+    {
+      ret = bpf_open_live (interface, filter);
+    }
+  else
+    {
+      if (pcap_findalldevs (&alldevsp, errbuf) == -1)
+        g_message ("Error for pcap_findalldevs(): %s", errbuf);
+      if (alldevsp != NULL)
+        /* get first device in list */
+        interface = g_strdup (alldevsp->name);
+      ret = bpf_open_live (interface, filter);
+      pcap_freealldevs (alldevsp);
+    }
 
   if (free_filter != 0)
     g_free (filter);
+
+  g_free (interface);
 
   return ret;
 }
@@ -148,6 +162,7 @@ init_v6_capture_device (struct in6_addr src, struct in6_addr dest, char *filter)
   int free_filter = 0;
   char name[INET6_ADDRSTRLEN];
   char errbuf[PCAP_ERRBUF_SIZE];
+  pcap_if_t *alldevsp = NULL; /* list of capture devices */
 
   a_src = g_strdup (inet_ntop (AF_INET6, &src, name, INET6_ADDRSTRLEN));
   a_dst = g_strdup (inet_ntop (AF_INET6, &dest, name, INET6_ADDRSTRLEN));
@@ -172,12 +187,25 @@ init_v6_capture_device (struct in6_addr src, struct in6_addr dest, char *filter)
   g_free (a_dst);
   g_free (a_src);
 
-  if ((interface = v6_routethrough (&src, &dest))
-      || (interface = pcap_lookupdev (errbuf)))
-    ret = bpf_open_live (interface, filter);
+  if ((interface = v6_routethrough (&src, &dest)))
+    {
+      ret = bpf_open_live (interface, filter);
+    }
+  else
+    {
+      if (pcap_findalldevs (&alldevsp, errbuf) == -1)
+        g_message ("Error for pcap_findalldevs(): %s", errbuf);
+      if (alldevsp != NULL)
+        /* get first device in list */
+        interface = g_strdup (alldevsp->name);
+      ret = bpf_open_live (interface, filter);
+      pcap_freealldevs (alldevsp);
+    }
 
   if (free_filter != 0)
     g_free (filter);
+
+  g_free (interface);
 
   return ret;
 }

--- a/nasl/nasl_packet_forgery.c
+++ b/nasl/nasl_packet_forgery.c
@@ -1509,6 +1509,7 @@ nasl_pcap_next (lex_ctxt *lexic)
   struct ip *ret = NULL;
   struct ip6_hdr *ret6 = NULL;
   char *filter = get_str_var_by_name (lexic, "pcap_filter");
+  pcap_if_t *alldevsp = NULL; /* list of capture devices */
   int timeout = get_int_var_by_name (lexic, "timeout", 5);
   tree_cell *retc;
   int sz;
@@ -1536,13 +1537,22 @@ nasl_pcap_next (lex_ctxt *lexic)
           interface = v6_routethrough (dst, &src);
         }
       if (interface == NULL)
-        interface = pcap_lookupdev (errbuf);
+        {
+          if (pcap_findalldevs (&alldevsp, errbuf) == -1)
+            g_message ("Error for pcap_findalldevs(): %s", errbuf);
+          if (alldevsp != NULL)
+            /* get first device in list */
+            interface = g_strdup (alldevsp->name);
+          pcap_freealldevs (alldevsp);
+        }
     }
 
   if (interface != NULL)
     {
       bpf = bpf_open_live (interface, filter);
     }
+  g_free (interface);
+
   if (bpf < 0)
     {
       nasl_perror (lexic, "pcap_next: Could not get a bpf\n");
@@ -1640,6 +1650,7 @@ nasl_send_capture (lex_ctxt *lexic)
   struct ip *ret = NULL;
   struct ip6_hdr *ret6 = NULL;
   char *filter = get_str_var_by_name (lexic, "pcap_filter");
+  pcap_if_t *alldevsp = NULL; /* list of capture devices */
   int timeout = get_int_var_by_name (lexic, "timeout", 5);
   tree_cell *retc;
   int sz;
@@ -1666,11 +1677,19 @@ nasl_send_capture (lex_ctxt *lexic)
           interface = v6_routethrough (dst, &src);
         }
       if (interface == NULL)
-        interface = pcap_lookupdev (errbuf);
+        {
+          if (pcap_findalldevs (&alldevsp, errbuf) == -1)
+            g_message ("Error for pcap_findalldevs(): %s", errbuf);
+          if (alldevsp != NULL)
+            /* get first device in list */
+            interface = g_strdup (alldevsp->name);
+          pcap_freealldevs (alldevsp);
+        }
     }
 
   if (interface != NULL)
     bpf = bpf_open_live (interface, filter);
+  g_free (interface);
 
   if (bpf < 0)
     {

